### PR TITLE
Issues & verifications when viewing / editing reports (Not report results) Fixes

### LINKED ISSUE #13: Issues & verifications when viewing / editing reports (Not report results)
Scout allows advanced reports to be created, viewed and edited for both computer and mobile devices. While they are saved as separate reports, the UI that they use is fundamentally the same. This means that bugs found in mobile device reports are likely found in computer reports as well. The database table that holds reports is called `reports` which holds general information about the reports, while `reports_line_items` stores information about each line item, in it's parent report. This table stores thing like the order in the report, criteria information, etc.

We have identified several issues when viewing and editing reports that need to be fixed. All of the report client side code lives in `/app/js/scripts.js`. By viewing reports we mean the report definition, not the query results of this application. 

**Tasks:**
1. It appears and extra line item is always added when viewing or editing an existing report. This could just be left around from generating the UI the first time. Either way, identify why this is happening and fix it. Verify the correct amount of saved line items are loaded in for both single line item and multiple line item reports. 
2. Sometimes the criteria and field to select field names aren't loaded into the report UI view properly. Figure out why this is happening and fix it. It may due to some async issues and how the html is generated. Make sure the report line items are being returned from the scout API properly, and then that the scout UI is loading them into the report view correct. 
3. The update existing report functionality appears to be broken. When editing an existing report you should be able to save it and it's updated in the database, however that's not the case. It was working in the past, so most of the functionality should be there on both the client and server, however there is an issue preventing the saving of reports right now. Investigate why and fix it. 

**PR Ready When:**
1. All three of the bugs are fixed, tested and verified. Viewing and editing existing reports should work in it's entirety.

### PR DESCRIPTION
**Pull request by alex-stout**
Resolves #13 

- [x] Fix extra line item showing up when editing or viewing a report. 
- [X] Fix update report. It was two issues:
    1. The field to select wasn't showing up so it wouldn't allow the `PUT`
    2. When the put was made the server wasn't handling the parenthesis as expected. I could've changed either the frontend or the server. I decided to change the server in order to prevent any issues in the future if there's bots connected to the API.
- [x] Fix criteria and field to select inputs not showing up

### CODE REVIEW

File: app/js/scripts.js [hunk 1]
```diff
@@ -1525,7 +1525,11 @@ function changeView(newView){
 }
 
 function changeReportView(deviceType, operation){
-  reloadReportPane(true);
+  if (operation == 'edit' || operation == 'view') {
+    reloadReportPane(false);
```
Comment (alex-stout) [inline app/js/scripts.js]: The true or false here determines if it's going to generate the new line item.

File: app/js/scripts.js [hunk 2]
```diff
@@ -137,6 +137,7 @@ function editReportById(reportId){
     changeReportView('computer', 'edit');
     $("#report-name-field").html(reportObject.name);
     $("#new-report-name").val(reportObject.name);
+    $("#existing-report-id").val(reportId)
```
Comment (alex-stout) [inline app/js/scripts.js]: This wasn't happening before and is required to update the report. It was thinking that it was a new report. That's why it was duplicating the report in the DB when it worked.

File: api/models/reports.js [hunk 3]
```diff
@@ -152,12 +152,16 @@ exports.updateReportById = function(reportObject, reportId){
 exports.updateReportLineItem = function(lineItem, order, reportId){
   //Set some meta fields
   lineItem.report_id = reportId;
-  //Check if we need to replace a parenthesis with a boolean value
+  // replace value with boolean to match DB schema
```
Comment (alex-stout) [inline api/models/reports.js]: Now catches anything other than a `(` or a `)` and makes it false in the DB.

**Comment by alex-stout** [comment]
There's an issue with the extra line item still showing up but it isn't due to the original problem.

Now it's because updating an existing report by removing a line item doesn't remove it from the DB. I'm working on a fix for that case.

**Comment by alex-stout** [comment]
It should be good to go now. The fix I implemented checks if there are less line items in the new one compared to the existing one. If there's less, it deletes the current ones and makes new ones for the new ones since the order will be different and not easily determined.

I also changed how the line items are queried and returned from the API. I made it so the line items are returned in ascending order by `item_order`.

**Comment by alex-stout** [comment]
Also should resolve #12

File: api/controllers/reports.js [hunk 4]
```diff
@@ -69,6 +69,7 @@ reports.get('/id/:reportId', function(req,res){
   //Get the report and it's line items from the database
   report.getReportById(req.params.reportId)
   .then(function(report){
+    console.log(report)
```
Comment (jacobschultz) [inline api/controllers/reports.js]: Do we need this console log?

Comment (alex-stout) [reply api/controllers/reports.js]: Nope.

**Comment by jacobschultz** [comment]
This looks pretty good after testing it - only found one thing. 

When you click the blue 'view icon' in the report list table, the 'Run report' button seems like it's trying to save it, the report name also isn't loaded. (The one with the eye icon)

I also feel like we should either disable the input fields in this view, or remove that button from the table completely. I don't know how much utility there is in this view now that I look again. If you want to run the report, you can use that button, and if you actually want to view the report criteria, you can just load up the edit view. 

What do you think?

**Comment by alex-stout** [comment]
I think you're right. It really doesn't have much functionality. I'll remove it.

**Comment by jacobschultz** [comment]
If there is any cleanup in the UI as far as removing code for viewing, feel free to do that too!

**Comment by alex-stout** [comment]
I created this report:
Screen Shot 2019-12-05 at 9 14 32 PM-crunch

And got this output (The reason why `model number` doesn't have anything is because our test environment doesn't have that field populated) : 
Screen Shot 2019-12-05 at 8 58 09 PM-crunch

It looks like everything is working for this test case.

**Pull request merged**